pkg/sentry/platform/ring0/defs_amd64.go

```diff
@@ -72,19 +72,59 @@ type KernelOpts struct {
 	PageTables *pagetables.PageTables
 }
 
+// Currently, KVM_MAX_VCPUS is 288, which has been being increased slowly.
+// So 512 for _RING0_MAX_CPUS is enough for a while.
+const _RING0_MAX_CPUS = 512
+
+type cpuEntryArea struct {
+	alignStart [usermem.PageSize]byte
+
+	// cpuEntries is array of kernelEntry for all cpus
+	cpuEntries [_RING0_MAX_CPUS]kernelEntry
+
+	// globalIDT is our set of interrupt gates.
+	globalIDT idt64
+
+	alignEnd [usermem.PageSize]byte
+}
+
 // KernelArchState contains architecture-specific state.
 type KernelArchState struct {
 	KernelOpts
 
-	// globalIDT is our set of interrupt gates.
-	globalIDT idt64
+	*cpuEntryArea
+	CpuEntryAreaStart, CpuEntryAreaEnd uintptr
 }
 
-// CPUArchState contains CPU-specific arch state.
-type CPUArchState struct {
+// kernelEntry contains minimal CPU-specific arch state
+// that can be mapped at the upper of the address space.
+// Malicious APP might steal info from it via CPU bugs.
+type kernelEntry struct {
 	// stack is the stack used for interrupts on this CPU.
 	stack [256]byte
 
+	// scratch space for temporary usage.
```

Comment:
Should the below be uintptr? uint64 also seems fine, just asking.

Reply:
stackTop is initiated via c.StackTop() which returns uint64.
So stackTop is defined as uint64, so as the scratch.